Write to Firehose

nathanleiby opened the pull request:
Issue: https://clever.atlassian.net/browse/INFRA-2209

Much of this is copied or built upon the SampleRecordProcessor from
https://github.com/clever/amazon-kinesis-client-go

The `firehose` package is based on our existing heka-clever-plugin
https://github.com/Clever/heka-clever-plugins/blob/master/firehose_output.go
and uses the same `batcher` to batch messages before sending.

Some areas to improve (mostly in upcoming PRs) include:

- (_fixed_) configurability (AWS creds, stream to read from, firehose to write to)
- (_won't fix in this PR_) there's no pre-processing / decoding logic
- checkpointing is over simplified. if the container crashed, it's possible things would be checkpointed that hadn't yet been flushed to Redshift
- (_fixed_) The Dockerfile is currently not very optimized (it sets up Go
internally and does the build there, moreover it runs the commands from
the Makefile). This was the quickest way for me to bootstrap, but should
be improved later.

nathanleiby committed:
first take: FirehoseWriter, consumer cmd, and Dockerfile

Much of this is copied or built upon the SampleRecordProcessor from
https://github.com/clever/amazon-kinesis-client-go

The `firehose` package is based on our existing heka-clever-plugin
https://github.com/Clever/heka-clever-plugins/blob/master/firehose_output.go
and uses the same `batcher` to batch messages before sending.

Note: The Dockerfile is currently not very optimized (it sets up Go
internally and does the build there, moreover it runs the commands from
the Makefile). This was the quickest way for me to bootstrap, but should
be improved later.

mohit commented on launch/kinesis-to-firehose.yml:
```diff
@@ -1,10 +1,10 @@
 resources:
-  cpu: 0.1
-  max_mem: 0.1
+  cpu: 1.0
```
should we start low here or is this is result of testing?

nathanleiby replied:
This is just ballpark numbers, but I'm pretty sure it'll be more than this.

I'll monitor and circle back on this.

nathanleiby committed:
Dockerfile: compile binary outside dockerfile

nathanleiby committed:
Makefile: allow passing env vars to consumer.properties

Run `make consumer_properties` to generate a consumer.properties file.

This will use the values from env vars:

- `KINESIS_STREAM_NAME` - name of stream
- `KINESIS_AWS_REGION` - AWS region where stream is running

nathanleiby committed:
firehose: configure region+stream via env vars

nathanleiby committed:
make env var names consistent, add to launch.yml

nathanleiby committed:
set app name based on _DEPLOY_ENV

mohit commented on Makefile:
```diff
@@ -52,10 +52,19 @@ install_deps: $(GOPATH)/bin/glide
+# sets the `applicationName` in KCL consume.properties.
+# we want a dif application in Dev and Prod, so that they don't conflict
+# (e.g. app name determines the DynamoDB table that KCL uses to coordinate)
+APPLICATION_NAME ?= kinesis-to-firehose-dev
```
why not just add them to `config` directly rather than relying on an internal `env_var`. We could just set `KINESIS_CONSUMER_NAME`?

nathanleiby replied:
(Same as https://github.com/Clever/kinesis-to-firehose/pull/1#discussion_r107053671)

nathanleiby commented:
I've verified that `ark start --local` is now working (pointed at config branch via `ark start --local -s INFRA-2209-kinesis-to-firehose`)

nathanleiby commented:
Related Ark Config: https://github.com/Clever/ark-config/pull/303

nathanleiby commented:
Update on areas to improve:

- (_fixed_ in 1f7af91 , 86fa7c7 ,  b6d8d2f , b649844) configurability (AWS creds, stream to read from, firehose to write to)
- (_won't fix in this PR_) there's no pre-processing / decoding logic
- (_looking into it, but probably won't fix in this PR_) checkpointing is over simplified. if the container crashed, it's possible things would be checkpointed that hadn't yet been flushed to Redshift
- (_fixed_ in ba1fbb0) The Dockerfile is currently not very optimized (it sets up Go
internally and does the build there, moreover it runs the commands from
the Makefile). This was the quickest way for me to bootstrap, but should
be improved later.

mohit commented on Dockerfile:
```diff
@@ -1 +1,12 @@
-FROM gliderlabs/alpine:3.3
+FROM openjdk:7
```
could use `openjdk:7-jre-alpine` or `openjdk:7-jre` (for debian). JDK is only required when building java apps, JRE includes only the runtime.

nathanleiby replied:
updated to use jre (debian) 71f9b4d280e415eee69fa03758fa802fffc2668d

mohit commented on Makefile:
```diff
@@ -1,13 +1,76 @@
+# sets the `applicationName` in KCL consume.properties.
+# we want a dif application in Dev and Prod, so that they don't conflict
+# (e.g. app name determines the DynamoDB table that KCL uses to coordinate)
+APPLICATION_NAME ?= kinesis-to-firehose-dev
```
I'd just default to use _KINESIS_APPLICATION_NAME_ defined in config. It's simpler than having it switch based on `_DEPLOY_ENV`

nathanleiby replied:
excellent point! updated to pass this as an env var and will update ark-config

b9bf9c10576ccdab4037a64f4f34023e895fa215

mohit commented on Dockerfile:
```diff
@@ -1 +1,12 @@
+# build
+ADD ./build/consumer ./build/consumer
+
+CMD ["make", "kinesis_consumer"]
```
have something here that makes it clear this is a `run` :)

nathanleiby replied:
Renamed to `run_kinesis_consumer` 31ceae0

mohit commented on cmd/consumer/main.go:
```diff
@@ -0,0 +1,175 @@
+
+func (rp *RecordProcessor) checkpoint(checkpointer kcl.Checkpointer, sequenceNumber string, subSequenceNumber int) {
+	for n := 0; n < rp.checkpointRetries; n++ {
+		if err := checkpointer.Checkpoint(sequenceNumber, subSequenceNumber); err == nil {
```
if you're `break`ing there should be no need for the `else`.

nathanleiby replied:
Refactored `checkpoint` logic 6d5b02d

mohit commented on cmd/consumer/main.go:
```diff
@@ -0,0 +1,175 @@
+	return nil
+}
+
+// func base64decode(
```
derp?

nathanleiby replied:
removed leftover commented code 78971247f30d17d6ded1f8a581e27ab7b9ed5658

mohit commented on cmd/consumer/main.go:
```diff
@@ -0,0 +1,175 @@
+		fmt.Fprintf(os.Stderr, "Was told to terminate, will attempt to checkpoint.\n")
+		rp.firehoseWriter.FlushAll()
+		rp.checkpoint(checkpointer, "", 0)
+	} else {
```
Never returns `error`. Also seems like this would be easier to read as:

```golang
if reason == "TERMINATE" {
...

return nil
}

fmt.Fprintf(os.Stderr, "Shutting down due to failover. Will not checkpoint.\n")
return nil
```

Also should we print the `reason` if it's not `TERMINATE`

fmt.Fprintf(os.Stderr, "Shutting down due to failover. Reason: %s. Will not checkpoint.\n", reason)

nathanleiby replied:
This fun implements the `kcl.RecordProcessor` interface, so it needs to return `error` for that reason.

I'll update to print out a clearer error!

nathanleiby replied:
Updated to print clearer error for Shutdown 39191b3

mohit commented on cmd/consumer/main.go:
```diff
@@ -0,0 +1,175 @@
+var streamName string
+var awsRegion string
+
+func main() {
```
pull this into `cmd/main.go` and then have the `cmd/consumer/consumer.go` in a different package?

nathanleiby replied:
I'll skip this refactor for now, but agreed if this file gets too complex, separating out could be good!

mohit commented on firehose/firehose_kinesis.go:
```diff
@@ -0,0 +1,125 @@
+	return nil
+}
+
+func convertToFields(msg string) map[string]interface{} {
```
could we just remove this stuff about Fields since we are not using it here?

nathanleiby replied:
removed convertToFields for now. will revisit when we add more decoding logic 8bef26211cce095b7b7d3bb5c6b8f40f46bfecfc

mohit commented on cmd/consumer/main.go:
```diff
@@ -0,0 +1,175 @@
+		}
+
+		// Handle checkpointing
+		// TODO: How to handle the difference between ProcessMessage (sent to FirehoseOutput) vs successfully sent?
```
In this code we should probably be only be concerned with successfully prepared to send to `Firehose` to begin with imo. Basically let's say we have `1000 incoming messages` we should say we parsed `990 messages to send to firehose`. If the issue is with the firehose endpoints failing (or rejecting messges) we can setup an alert on `consumer sent/sec` - `firehose recieved/sec`. 

What do you think?

nathanleiby replied:
My main concern here is if the app crashes.

As soon as a message is handed off to the `batcher`, it's considered processed. However, if they batcher holds onto it (say for 1 minute), it's possible that the app could crash during this time. If a checkpoint also occurs, KCL will consider those messages processed, even though our `batcher` never even attempted to send them.

One idea I considered is whether we might pursue something like "deliver" (handed to batcher) vs "ack" (batcher confirms messages was processed successfully)... here inspired by [RabbitMQ confirmations](https://www.rabbitmq.com/confirms.html)

nathanleiby replied:
I also want to make sure the graceful shutdown flow (flush everything, then checkpoint, then stop) works.

mohit commented on cmd/consumer/main.go:
```diff
@@ -0,0 +1,175 @@
+		rp.lastCheckpoint = time.Now()
+
+		// Write status to file
+		err := appendToFile("/tmp/kcl_stderr", fmt.Sprintf("%s -- %+v\n", rp.shardID, rp.firehoseWriter.Status()))
```
strong suggestion to take this `filename` an a CLI arg (preferred) or env var

nathanleiby replied:
I was unsure if we could invoke the consumer binary with any flags, so I went with env var approach d5467a7c375aa88c8943a5bf282f889f096c88c6

mohit reviewed (commented):
still looking through the checkpointing conditions. Just some nits and general comments.

nathanleiby committed:
Makefile: set KINESIS_APPLICATION_NAME via env var

previously, was implicitly set using `_DEPLOY_ENV`, an env var injected
at runtime by Catapult.

nathanleiby committed:
Dockerfile: use openjdk:7-jre, since only runtime needed

nathanleiby committed:
cleanup derp

nathanleiby committed:
firehose: remove convertToFields

not relevant right now. can add back something similar once we have
decoding logic

nathanleiby committed:
s/kinesis_consumer/run_kinesis_consumer

nathanleiby committed:
cmd: made LOG_FILE configureable

nathanleiby committed:
cmd: print reason for Shutdown

nathanleiby committed:
cmd: refactor checkpointing logic

nathanleiby committed:
launch: add KINESIS_APPLICATION_NAME to env

nathanleiby committed:
launch: add LOG_FILE to env

nathanleiby committed:
README: clarify usage

Also remove default LOG_FILE from Makefile, since it wasn't being set
correctly in env anyway.

nathanleiby committed:
checkpoint: give up after n attempts; dont sleep one last time

nathanleiby commented:
@mohit I believe I've responded to all feedback.

mohit commented on cmd/consumer/main.go:
```diff
@@ -0,0 +1,179 @@
+
+func (rp *RecordProcessor) Initialize(shardID string) error {
+	rp.shardID = shardID
+	rp.lastCheckpoint = time.Now()
```
a comment for why this is set to Now? or just an explanation for me?

nathanleiby replied:
This follows the approach used in `awslabs` example record processors https://github.com/awslabs/amazon-kinesis-client-python/blob/master/samples/sample_kclpy_app.py#L50

We're just using it so that we don't attempt to Checkpoint too frequently.

```go
if time.Now().Sub(rp.lastCheckpoint) > rp.checkpointFreq {
   // ... checkpoint
}
```

mohit commented on cmd/consumer/main.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+func (rp *RecordProcessor) shouldUpdateSequence(sequenceNumber *big.Int, subSequenceNumber int) bool {
+	return rp.largestSeq == nil || sequenceNumber.Cmp(rp.largestSeq) == 1 ||
```
what does this mean -- comment?

nathanleiby replied:
Added a comment 2ba6b8716600248bdf78a9057a80a5c9ee00b0a8


also, here's more info on how sequencing within the kinesis stream works: http://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-consumer-deaggregation.html

nathanleiby committed:
cmd: more idiomatic Go error handling

nathanleiby committed:
cmd: comment for shouldUpdateSequence
